lib/THC/THCTensorMode.cuh

```diff
@@ -0,0 +1,292 @@
+  // assumes scan syncs at the end
+
+  // Next, we reinterpret the ubpmem buffer as pairs of unsigned integers (i.e. we treat the
+  // boolean flag regions as integers). We initialize these to represent indices, and we'll call
+  // this buffer I
+  struct ModeUnsignedPair *uupmem = reinterpret_cast<struct ModeUnsignedPair *>(ubpmem);
+  uupmem[tidx].index = tidx;
+  uupmem[stidx].index = stidx;
+  __syncthreads(); // barrier for index initialization
+
+  // At this point, we need to find the maximum element in lengths buffer B.
+  // This element will represent the count (-1) of the mode. Because of the
+  // way we have set up the problem, the index where this mode occurs will
+  // also be the location of the mode value in the sorted array, e.g.
+  //
+  // smem = [0, 0, 1, 1, 1, 2]
+  // C    = [0, 1, 0, 1, 2, 0]
+  // I    = [0, 1, 2, 3, 4, 5]
+  //                     ^
+  //                     maximum value, also aligned with mode = 1
+  //
+  // We perform a block wide max-reduction of the C buffer, but we also need the
+  // indices to come along with it, so we utilize the uupmem construction.
+  //
+  // Loop 1 (Power2Size = offset = 4):
+  //
+  // (0, 4) --> C[4] = 2 > C[0] = 0, so update C[0], I[0]
+  // (1, 5) --> C[5] = 0 <= C[1] = 1, do nothing
+  //
+  // Now:         0  1  2  3  4  5
+  //      C = [2, 1, 0, 1, 2, 0]
+  //      I = [4, 1, 2, 3, 4, 5]
+  //
+  // Loop 2 (offset = 2)
+  //
+  // (0, 2) --> C[2] == 0 <= C[0] = 2, do nothing
+  // (1, 3) --> C[3] == 1 <= C[1] = 1, do nothing
+  //
+  // Now:         0  1  2  3  4  5
+  //      C = [2, 1, 2, 1, 2, 0]
+  //      I = [4, 1, 4, 3, 4, 5]
+  //
+  // Loop 3 (offset = 1)
+  //
+  // (0, 1) --> C[1] == 1 <= C[0] = 2, do nothing
+  //
+  // So at the end at C[0] we have the maximum count = 2, and the
+  // corresponding index I[0] = 4
+
+  struct ModeUnsignedPair max = {0, 0};
+  max = reduceBlockN<struct ModeUnsignedPair, MaxReduceOp<struct ModeUnsignedPair>, 2, true>
+    (uupmem, &uupmem[tidx * 2], sliceSize, MaxReduceOp<struct ModeUnsignedPair>(), max);
+
+  // Store the mode in shared memory for use in finding the mode in the input slice
+  __shared__ T  mode;
+
+  // Given the above constraints, the mode is the value at the reduced index in the
+  // original sorted element buffer
+  if (tidx == 0) {
+    mode = smem[max.index];
+  }
+  __syncthreads(); // broadcast mode
+
+  // Finally, we need to find the "an" index of the mode in the input Tensor. The API does
+  // not constrain which index we pick, so it can be any of the indices that contain the mode.
+  // We will do a reduction to find the index. We go back to using the (index, flag) buffer
+  // arrangment. First, we mark indices that are equal to the mode, i.e B[i] = true if
+  // input[i] == mode, and initialize C[i] to be the index
+  if (tidx < sliceSize) {
+    ubpmem[tidx].flag = THCNumerics<T>::eq(input[linearOffset + tidx], mode);
```

Comment:
Same as above: why store this in shared memory? The point of reduceBlockN is that the input data can be held in registers